src/Stratis.Bitcoin.Features.MemoryPool/BlocksDisconnectedSignaled.cs

```diff
@@ -65,6 +75,28 @@ private async Task AddBackToMempoolAsync(Block block)
             }).ConfigureAwait(false);
         }
 
+        /// <summary>
+        /// If there are any transactions in the mempool that depend on transactions no longer in the chain, remove them.
+        /// </summary>
+        private async Task RemoveInvalidTransactionsAsync(Block block)
+        {
+            // Invalid transactions would have spent the coinbase. The other transactions can be put back into the mempool and be fine.
+            uint256 coinbaseId = block.Transactions[0].GetHash();
+
+            await this.mempoolLock.WriteAsync(async () =>
+            {
+                foreach (TxMempoolEntry mempoolEntry in this.mempool.MapTx.SpendsCoinbase.ToList())
```

Comment:
I agree that this is important logic that needs to be added, but on further consideration I think it is incomplete.

Consider the case where we are on fork A and reorg to fork B. Now all of the transactions on fork A that are beyond the fork point need to be re-evaluated. This is particularly true in the case of multi-block reorgs: a transaction may be put back into the pool when the tip is disconnected, then its own input(s) maybe be removed when (tip-1) is disconnected, etc.

At minimum the following would need to be re-evaluated, in my view:
a. Existence of inputs (they may have been reorged away; for example the coinbase, such as in your PR)
b. Coinbase/stake maturity requirements (it is possible that fork B is shorter than A such that these requirements are no longer met)
c. Similarly to (b), locktime requirements
d. Possibly others

For now this PR can be merged standalone, but I would suggest adding a TODO as a reminder to do the rest.

Reply:
Thanks, made the changes as suggested.